src/iosrv/srv.c

```diff
@@ -770,23 +806,42 @@ dss_xstreams_init()
 	D_DEBUG(DB_TRACE, "%d cores detected, starting %d main xstreams\n",
 		dss_core_nr, dss_tgt_nr);
 
-	for (i = 0; i < DSS_XS_NR_TOTAL; i++) {
-		hwloc_obj_t	obj;
+	xstream_data.xd_xs_nr = DSS_XS_NR_TOTAL;
+	/* start system service XS */
+	for (i = 0; i < dss_sys_xs_nr; i++) {
+		xs_id = i;
+		rc = dss_start_xs_id(xs_id);
+		if (rc)
+			D_GOTO(out, rc);
+	}
 
-		obj = hwloc_get_obj_by_depth(dss_topo, dss_core_depth,
-					     i % dss_core_nr);
-		if (obj == NULL) {
-			D_ERROR("Null core returned by hwloc\n");
-			D_GOTO(failed, rc = -DER_INVAL);
-		}
+	/* start main IO service XS */
+	for (i = 0; i < dss_tgt_nr; i++) {
+		xs_id = DSS_MAIN_XS_ID(i);
+		rc = dss_start_xs_id(xs_id);
+		if (rc)
+			D_GOTO(out, rc);
+	}
 
-		rc = dss_start_one_xstream(obj->allowed_cpuset, i);
+	/* start offload XS if any */
+	if (dss_tgt_offload_xs_nr == 0)
+		D_GOTO(out, rc);
+	for (i = 0; i < dss_tgt_nr; i++) {
+		xs_id = DSS_MAIN_XS_ID(i) + 1;
+		rc = dss_start_xs_id(xs_id);
+		if (rc)
+			D_GOTO(out, rc);
+		if (dss_tgt_offload_xs_nr == 1)
+			continue;
+		xs_id++;
+		rc = dss_start_xs_id(xs_id);
```

Comment:
it seems you can use another for by dss_tgt_offload_xs_nr, then you do not need the previous if as well?
for (j = 0; j < dss_tgt_offload_xs_nr; j++) {
            ...
}

Reply:
ok